doc/apcctrl.8
.\" manual page [] for apcctrl version 3.14.6
.Dd January 10, 2009
.Dt APCCTRL 8
.Os apcctrl v3.14.6
.Sh NAME
.Nm apcctrl 
.Nd a daemon for controlling most APC UPSes
.Sh SYNOPSIS
.Nm
.Op Fl b 
.Op Fl d Ar level | Fl -debug Ar level
.Op Fl f Ar file | Fl -config-file Ar file
.Op Fl P Ar file | Fl -pid-file Ar file
.Op Fl p | -kill-on-powerfail | t | -term-on-powerfail
.Op Fl R
.Pp
.Nm 
.Op Fl k | -killpower | -hibernate | o | -power-off 
.Pp
.Nm
.Op Fl h | -help
.Pp
.Nm
.Op Fl V | -version
.Pp
.Sh DESCRIPTION
.Pp
The
.Nm 
daemon controls the operation of most American Power Conversion Corp 
(APC) UPSes.  During a power failure,
.Nm
informs users about the loss of utility power and that a shutdown may occur.  
If utility power is not restored, a system shutdown will follow when the 
battery is exhausted, a specified timeout expires, a specified battery 
charge percentage is reached, or a specified battery runtime (based on 
internal UPS calculations and determined by power consumption rates)
expires.  If the utility power is restored before one of the these 
shutdown conditions is met, 
.Nm
will inform users of this and the shutdown will generally be cancelled.
Refer to the Implementation Notes section below for situations in which 
the shutdown may not be cancelled.
.Pp
The meaning of the command line options is as follows:
.Bl -tag -width Fl
.It Fl b
Run in the foreground, do not detach and become a daemon.
.It Fl d Ar level Fl -debug Ar level
Set debugging output level where level is a number greater than zero.
.It Fl f Ar file Fl -config-file Ar file
Load the specified configuration file. 
The default configuration file is 
.Pa /etc/apcctrl/apcctrl.conf .
It must be changed when running more than one copy of
.Nm
on the same computer to control multiple UPSes.
.It Fl k | Fl -killpower | -hibernate
Power down the UPS in hibernate mode after a 10 second delay. This 
option is normally only used by 
.Nm
itself to power down the UPS after a system shutdown has almost completed. 
In hibernate mode, the UPS will again supply power to the system when the 
utility power returns.
.It Fl o | -power-off
Power down UPS completely. The UPS will not supply power to the system
when the utility power returns.
.It Fl P Ar file Fl -pid-file Ar file
Create the specified process ID file. The default is
.Pa /var/run/apcctrl.pid .
It must be changed when running more than one copy of
.Nm
on the same computer to control multiple UPSes.
.It Fl p -kill-on-powerfail
.Nm
commands the UPS to power down in hibernate mode just before
.Nm
starts the system shutdown. This relies on the grace shutdown delay 
of a Smart-UPS being long enough to allow the system to shutdown 
completely before the UPS shuts off the power to the system and goes 
into hibernate mode. This shutdown grace delay is a programmable
value stored in a Smart-UPS EEPROM which can be changed using
.Xr apctest 8 .
In hibernate mode, the UPS will again supply power 
to the system when the utility power returns. Refer to
.Xr apcctrl.conf 5
for an alternative method using the KILLDELAY configuration 
directive and the only method available when using a Back-UPS or other 
UPS operating in simple signalling mode.
.It Fl t -term-on-powerfail
.Nm
exits immediately when it issues the system shutdown command. 
This behaviour is useful for those systems where it is not possible 
to insert 
.Nm
commands in a halt script to issue the killpower command. 
Without this option, 
.Nm
will wait for the SIGTERM signal from the system shutdown process
before exiting.
.It Fl R
Put a UPS which runs in smart signalling mode by default (eg a Smart-UPS)
into simple signalling mode.
.It Fl v -version
Prints the
.Nm
version number and the program help.
.It Fl h  -help
Prints the program help.
.Pp
.El
The
.Nm
daemon supports a networking mode called Network 
Information Server  (NIS) -- not related  to Sun's NIS/YP -- 
in which the daemon serves status and event information to 
clients over a network. See the "Running The Network Information 
Server" section of the apcctrl manual for more information and 
configuration details on this mode.
.Pp
The 
.Nm
daemon also supports a Smart-UPS in SNMP mode provided an APC
Web/SNMP or APC PowerNet SNMP card is installed in the UPS's SmartSlot. 
For more information and configuration details on this mode, refer to the 
"Support for SNMP UPSes" section of the apcctrl manual.
.Pp
.Sh IMPLEMENTATION NOTES
.Pp
The shutdown is made by calls to the script
.Pa /etc/apcctrl/apccontrol
by the
.Nm
daemon.  Consequently, no changes to
.Pa /etc/inittab
are necessary on Linux as there is no communication between the 
daemon and the init(1) process.  Installation of the daemon modifies 
the halt script so that at the end of the shutdown process, the daemon
will be re-executed to power off the UPS in hibernate mode. 
.Pp
On some operating systems (eg FreeBSD) there is no halt script so apccontrol 
must be modified to cause the daemon to power off the UPS after a delay. 
Alternatives are to use the --kill-on-powerfail command on the
.Nm 
command line or refer to 
.Xr apcctrl.conf 5
for details of the KILLDELAY configuration directive. 
.Pp
.Sh CONFIGURATION
It will almost certainly be necessary to customise the configuration 
information in the
.Pa /etc/apcctrl/apcctrl.conf
configuration file to suit your configuration and desired UPS behaviour.
.Pp
For information on the configuration directives and the format of 
the configuration file, refer to 
.Xr apcctrl.conf 5 .
.Pp
.Sh EVENTS
.Pp
.Nm
generates events when certain conditions occur. 
These events are sent to the system log and, optionally,
to the temporary events file 
.Pa ( /var/log/apcctrl.events ) .
They also generate a call to  the
.Pa /etc/apcctrl/apccontrol
script which in turn will call any custom scripts placed in the
.Pa /etc/apcctrl
directory which may override apccontrol's default behaviour. For 
details of the events and customising apccontrol's behaviour, refer to
.Xr apccontrol 8 .
.Pp
.Sh DATA FILE FORMAT
.Pp
If the DATATIME configuration directive is set to non-zero, 
.Nm
will log a data record at the interval defined
by the DATATIME directive. This data record is in a format
similar to the APC PowerChute software data file format.
.Sh STATUS REPORT FORMAT
The status report output format is simple ASCII. Generally there
is a single piece of information on each line of output. The
content varies based on the model of UPS being used and, in
some cases, the firmware version. This status report is also
optionally written the the 
.Pa /etc/apcctrl/apcctrl.status
file. Refer to
.Xr apcaccess 8
for full details of the status report output.
.Sh FILES
.Pa /etc/apcctrl/apcctrl.conf 
default configuration file
.Pp
.Pa /var/run/apcctrl.pid 
default process ID file
.Pp
.Pa /var/log/apcctrl.status 
optional status file
.Pp
.Pa /var/log/apcctrl.events
default events file
.Pp
.Sh SEE ALSO
.Pp
.Xr apcctrl.conf 5 ,
.Xr apcaccess 8 ,
.Xr apccontrol 8 ,
.Xr apctest 8 .
.Pp
The HTML apcctrl manual installed on your system or available online at
http://www.apcctrl.org/
.Pp
.Sh AUTHORS
.Pp
.Ss This page
.An Trevor Roydhouse (current)
.An Andre M. Hedrick 
.An Christopher J. Reimer 
.Pp
.Ss Software
.An Adam Kropelin (current Project Manager and Code Maintainer)
.An Kern Sibbald (former Project Manager and Code Maintainer)
.An Riccardo Facchetti (former Project Manager and Code Maintainer)
.An Andre M. Hedrick (Project Founder and former Code Maintainer)
.Pp
.Ss Contributors
.Pp
An enormous number of users who have devoted their time and energy to 
this project -- thanks.


ChangeLog
/***************************************************************************/
/*                     ChangeLog of apcctrl 3.14.x                         */
/*                                                                         */
/*               Adam Kropelin <adam@kroptech.com>                         */
/*                                                                         */
/*               http://www.apcctrl.com/                                   */
/*               http://sourceforge.net/projects/apcctrl                   */
/***************************************************************************/

------------------------------------------------------------------------
r2314 | adk0212 | 2015-02-02 21:16:47 -0500 (Mon, 02 Feb 2015) | 2 lines

Correct some outdated information in the apctest section of the manual

------------------------------------------------------------------------
r2313 | adk0212 | 2015-02-02 20:47:30 -0500 (Mon, 02 Feb 2015) | 2 lines

Detect failure to modify RedHat halt script and display a message instead of failing the whole installation

------------------------------------------------------------------------
r2312 | adk0212 | 2015-01-31 11:38:09 -0500 (Sat, 31 Jan 2015) | 2 lines

Prep for 3.14.13 release

------------------------------------------------------------------------
r2311 | adk0212 | 2015-01-29 09:13:25 -0500 (Thu, 29 Jan 2015) | 5 lines

Place apcctrl lock file in /var/lock instead of /var/lock/subsys on Slackware
Slackware convention is to use /var/lock; /var/lock/subsys appears to be a Red Hat thing
Reported by Sebastian Arcus <s.arcus@open-t.co.uk>
Patch by Jonathan Woithe <jwoithe@atrad.com.au>

------------------------------------------------------------------------
r2310 | adk0212 | 2015-01-20 12:38:53 -0500 (Tue, 20 Jan 2015) | 3 lines

Add notes indicating that apcctrl must be restarted for config file changes to become active.
Although it might be considered "obvious", this wasn't actually written down anywhere that I could find.

------------------------------------------------------------------------
r2309 | adk0212 | 2015-01-20 12:15:04 -0500 (Tue, 20 Jan 2015) | 2 lines

Updated Windows USB driver signature, courtesy of Jernej Simončič <jernej|s@eternallybored.org>

------------------------------------------------------------------------
r2308 | adk0212 | 2015-01-19 21:45:44 -0500 (Mon, 19 Jan 2015) | 2 lines

Add debug to Tx side of modbus USB to match RS232

------------------------------------------------------------------------
r2307 | adk0212 | 2015-01-19 21:02:01 -0500 (Mon, 19 Jan 2015) | 2 lines

Remove unused a{m,re,c}alloc functions

------------------------------------------------------------------------
r2306 | adk0212 | 2015-01-19 20:58:51 -0500 (Mon, 19 Jan 2015) | 2 lines

Remove obsolete UPS list code

------------------------------------------------------------------------
r2305 | adk0212 | 2015-01-19 20:55:56 -0500 (Mon, 19 Jan 2015) | 2 lines

Remove obsolete references to myUPS

------------------------------------------------------------------------
r2304 | adk0212 | 2015-01-19 20:52:18 -0500 (Mon, 19 Jan 2015) | 2 lines

Remove obsolete file

------------------------------------------------------------------------
r2303 | adk0212 | 2015-01-19 20:48:49 -0500 (Mon, 19 Jan 2015) | 2 lines

Remove obsolete prototypes

------------------------------------------------------------------------
r2302 | adk0212 | 2015-01-19 20:38:19 -0500 (Mon, 19 Jan 2015) | 2 lines

Remove obsolete simple signaling cable info

------------------------------------------------------------------------
r2301 | adk0212 | 2015-01-19 19:09:42 -0500 (Mon, 19 Jan 2015) | 2 lines

Bump win32 usb driver date and rev

------------------------------------------------------------------------
r2300 | adk0212 | 2015-01-19 19:04:40 -0500 (Mon, 19 Jan 2015) | 2 lines

Fix typo

------------------------------------------------------------------------
r2299 | adk0212 | 2015-01-19 18:52:30 -0500 (Mon, 19 Jan 2015) | 2 lines

Fix typo

------------------------------------------------------------------------
r2298 | adk0212 | 2015-01-18 15:38:34 -0500 (Sun, 18 Jan 2015) | 2 lines

Add --enable-modbus-usb to Darwin build notes

------------------------------------------------------------------------
r2297 | adk0212 | 2015-01-18 11:41:26 -0500 (Sun, 18 Jan 2015) | 3 lines

Update supported USB VID/PID combinations in code and platform-specific drivers
- Includes several VID/PID tuples for future use as suggested by APC

------------------------------------------------------------------------
r2296 | adk0212 | 2015-01-12 11:56:25 -0500 (Mon, 12 Jan 2015) | 3 lines

Remove ancient unmaintained apcctrl.8 manpage from Debian platform directory.
The current manpage set for apcctrl is in doc/ and should be used by all platforms.

------------------------------------------------------------------------
r2295 | adk0212 | 2015-01-07 22:06:44 -0500 (Wed, 07 Jan 2015) | 2 lines

Fix handling of return value from USBDEVFS_CONNECT

------------------------------------------------------------------------
r2294 | adk0212 | 2015-01-07 21:31:27 -0500 (Wed, 07 Jan 2015) | 2 lines

Add MODBUS/USB info to apcctrl.conf comments

------------------------------------------------------------------------
r2293 | adk0212 | 2015-01-07 21:25:06 -0500 (Wed, 07 Jan 2015) | 2 lines

Add MODBUS USB setting info to apcctrl.conf(5) man page

------------------------------------------------------------------------
r2292 | adk0212 | 2015-01-07 20:54:03 -0500 (Wed, 07 Jan 2015) | 2 lines

Remove ETIMEDOUT workaround; this is no longer an issue

------------------------------------------------------------------------
r2291 | adk0212 | 2015-01-04 19:24:06 -0500 (Sun, 04 Jan 2015) | 2 lines

Update documentation for MODBUS USB.

------------------------------------------------------------------------
r2289 | adk0212 | 2015-01-04 15:41:14 -0500 (Sun, 04 Jan 2015) | 2 lines


src/apcctrl.c
/*
 * apcctrl.c
 *
 * APC UPS monitoring daemon.
 */

/*
 * Copyright (C) 1999-2005 Kern Sibbald
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of version 2 of the GNU General
 * Public License as published by the Free Software Foundation.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
 * General Public License for more details.
 *
 * You should have received a copy of the GNU General Public
 * License along with this program; if not, write to the Free
 * Software Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston,
 * MA 02110-1335, USA.
 */

#include "apc.h"

UPSINFO *core_ups = NULL;

static void daemon_start(void);

int pidcreated = 0;
extern int kill_on_powerfail;
extern FILE *trace_fd;
extern char *pidfile;

/*
 * The terminate function and trapping signals allows apcctrl
 * to exit and cleanly close logfiles, and reset the tty back
 * to its original settings. You may want to add this
 * to all configurations.  Of course, the file descriptors
 * must be global for this to work, I also set them to
 * NULL initially to allow terminate to determine whether
 * it should close them.
 */
void apcctrl_terminate(int sig)
{
   UPSINFO *ups = core_ups;

   if (sig != 0)
      log_event(ups, LOG_WARNING, "apcctrl exiting, signal %u", sig);

   clean_threads();
   clear_files();
   if (ups->driver)
      device_close(ups);
   delete_lockfile(ups);
   if (pidcreated)
      unlink(pidfile);
   log_event(ups, LOG_NOTICE, "apcctrl shutdown succeeded");
   destroy_ups(ups);
   closelog();
   _exit(0);
}

void apcctrl_error_cleanup(UPSINFO *ups)
{
   if (ups->driver)
      device_close(ups);
   delete_lockfile(ups);
   if (pidcreated)
      unlink(pidfile);
   clean_threads();
   log_event(ups, LOG_ERR, "apcctrl error shutdown completed");
   destroy_ups(ups);
   closelog();
   exit(1);
}

/*
 * Subroutine error_out prints FATAL ERROR with file,
 * line number, and the error message then cleans up 
 * and exits. It is normally called from the Error_abort
 * define, which inserts the file and line number.
 */
void apcctrl_error_out(const char *file, int line, const char *fmt, va_list arg_ptr)
{
   char buf[256];
   int i;

   asnprintf(buf, sizeof(buf),
      "apcctrl FATAL ERROR in %s at line %d\n", file, line);

   i = strlen(buf);
   avsnprintf((char *)&buf[i], sizeof(buf) - i, (char *)fmt, arg_ptr);

   fprintf(stderr, "%s", buf);
   log_event(core_ups, LOG_ERR, "%s", buf);
   apcctrl_error_cleanup(core_ups);     /* finish the work */
}

/*
 * apcctrlMain is called from win32/winmain.cpp
 * we need to eliminate "main" as an entry point,
 * otherwise, it interferes with the Windows
 * startup.
 */
#ifdef HAVE_MINGW
# define main apcctrlMain
#endif

/* Main program */
int main(int argc, char *argv[])
{
   UPSINFO *ups;
   int tmp_fd, i;

   /* Set specific error_* handlers. */
   error_out = apcctrl_error_out;

   /*
    * Default config file. If we set a config file in startup switches, it
    * will be re-filled by parse_options()
    */
   cfgfile = APCCONF;

   /*
    * Create fake stdout in order to circumvent problems
    * which occur if apcctrl doesn't have a valid stdout
    * Fix by Alexander Schremmer <alex at alexanderweb dot de>
    */
   tmp_fd = open("/dev/null", O_RDONLY);
   if (tmp_fd > 2) {
      close(tmp_fd);
   } else if (tmp_fd >= 0) {
      for (i = 1; tmp_fd + i <= 2; i++)
         dup2(tmp_fd, tmp_fd + i);
   }

   /*
    * If there's not one in libc, then we have to use our own version
    * which requires initialization.
    */
   strlcpy(argvalue, argv[0], sizeof(argvalue));

   ups = new_ups();                /* get new ups */
   if (!ups)
      Error_abort("%s: init_ipc failed.\n", argv[0]);

   init_ups_struct(ups);
   core_ups = ups;                 /* this is our core ups structure */

   /* parse_options is self messaging on errors, so we need only to exit() */
   if (parse_options(argc, argv))
      exit(1);

   Dmsg(10, "Options parsed.\n");

   if (show_version) {
      printf("apcctrl " APCCTRL_RELEASE " (" ADATE ") " APCCTRL_HOST "\n");
      exit(0);
   }

   /*
    * We open the log file early to be able to write to it
    * it at any time. However, please note that if the user
    * specifies a different facility in the config file
    * the log will be closed and reopened (see match_facility())
    * in apcconfig.c.  Any changes here should also be made
    * to the code in match_facility().
    */
   openlog("apcctrl", LOG_CONS | LOG_PID, ups->sysfac);

#ifndef DEBUG
   if ((getuid() != 0) && (geteuid() != 0))
      Error_abort("Needs super user privileges to run.\n");
#endif

   check_for_config(ups, cfgfile);
   Dmsg(10, "Config file %s processed.\n", cfgfile);

   /*
    * Disallow --kill-on-powerfail in conjunction with simple signaling
    * UPSes. Such UPSes have no shutdown grace period so using --kill-on-
    * powerfail would guarantee an unclean shutdown.
    */
   if (kill_on_powerfail && ups->mode.type == DUMB_UPS) {
      kill_on_powerfail = 0;
      log_event(ups, LOG_WARNING,
         "Ignoring --kill-on-powerfail since it is unsafe "
            "on Simple Signaling UPSes");
   }

   /* Attach the correct driver. */
   attach_driver(ups);
   if (ups->driver == NULL)
      Error_abort("apcctrl cannot continue without a valid driver.\n");

   ups->start_time = time(NULL);

   if (!hibernate_ups && !shutdown_ups && go_background) {
      daemon_start();

      /* Reopen log file, closed during becoming daemon */
      openlog("apcctrl", LOG_CONS | LOG_PID, ups->sysfac);
   }

   init_signals(apcctrl_terminate);

   /* Create temp events file if we are not doing a hibernate or shutdown */
   if (!hibernate_ups && !shutdown_ups && ups->eventfile[0] != 0) {
      ups->event_fd = open(ups->eventfile, O_RDWR | O_CREAT | O_APPEND | O_CLOEXEC, 0644);
      if (ups->event_fd < 0) {
         log_event(ups, LOG_WARNING, "Could not open events file %s: %s\n",
            ups->eventfile, strerror(errno));
      }
   }

   if (create_lockfile(ups) == LCKERROR) {
      Error_abort("Unable to create UPS lock file.\n"
                   "  If apcctrl or apctest is already running,\n"
                   "  please stop it and run this program again.\n");
   }

   make_pid_file();
   pidcreated = 1;

   if (hibernate_ups || shutdown_ups)
   {
      // If we're hibernating or shutting down the UPS, setup is a one-shot.
      // If it fails, we're toast; no retries
      if (!setup_device(ups))
         Error_abort("Unable to open UPS device for hibernate or shutdown");

      if (hibernate_ups)
         initiate_hibernate(ups);
      else
         initiate_shutdown(ups);

      apcctrl_terminate(0);
   }

   /*
    * From now ... we must _only_ start up threads!
    * No more unchecked writes to UPSINFO because the threads must rely
    * on write locks and up to date data in the shared structure.
    */

   /* Network status information server */
   if (ups->netstats) {
      start_thread(ups, do_server, "apcnis", argv[0]);
      Dmsg(10, "NIS thread started.\n");
   }

   log_event(ups, LOG_NOTICE,
      "apcctrl " APCCTRL_RELEASE " (" ADATE ") " APCCTRL_HOST " startup succeeded");

   /* main processing loop */
   do_device(ups);

   apcctrl_terminate(0);
   return 0;                       /* to keep compiler happy */
}

extern int debug_level;

/*
 * Initialize a daemon process completely detaching us from
 * any terminal processes.
 */
static void daemon_start(void)
{
#if !defined(HAVE_MINGW)
   int i, fd;
   pid_t cpid;
   mode_t oldmask;

#ifndef HAVE_QNX_OS
   if ((cpid = fork()) < 0)
      Error_abort("Cannot fork to become daemon\n");
   else if (cpid > 0)
      exit(0);                     /* parent exits */

   /* Child continues */
   setsid();                       /* become session leader */
#else
   if (procmgr_daemon(EXIT_SUCCESS, PROCMGR_DAEMON_NOCHDIR
                                    | PROCMGR_DAEMON_NOCLOSE
                                    | PROCMGR_DAEMON_NODEVNULL
                                    | PROCMGR_DAEMON_KEEPUMASK) == -1)
   {
      Error_abort("Couldn't become daemon\n");
   }
#endif   /* HAVE_QNX_OS */

   /* Call closelog() to close syslog file descriptor */
   closelog();

   /*
    * In the PRODUCTION system, we close ALL file descriptors unless
    * debugging is on then we leave stdin, stdout, and stderr open.
    * We also take care to leave the trace fd open if tracing is on.
    * Furthermore, if tracing we redirect stdout and stderr to the
    * trace log.
    */
   for (i=0; i<sysconf(_SC_OPEN_MAX); i++) {
      if (debug_level && i == STDIN_FILENO)
         continue;
      if (trace_fd && i == fileno(trace_fd))
         continue;
      if (debug_level && (i == STDOUT_FILENO || i == STDERR_FILENO)) {
         if (trace_fd)
            dup2(fileno(trace_fd), i);
         continue;
      }
      close(i);
   }

   /* move to root directory */
   chdir("/");

   /* 
    * Avoid creating files 0666 but don't override a
    * more restrictive umask set by the caller.
    */
   oldmask = umask(022);
   oldmask |= 022;
   umask(oldmask);

   /*
    * Make sure we have fd's 0, 1, 2 open
    * If we don't do this one of our sockets may open
    * there and if we then use stdout, it could
    * send total garbage to our socket.
    */
   fd = open("/dev/null", O_RDONLY);
   if (fd > 2) {
      close(fd);
   } else if (fd >= 0) {
      for (i = 1; fd + i <= 2; i++)
         dup2(fd, fd + i);
   }
#endif   /* HAVE_MINGW */
}
